### PR TITLE
[browser] logging and exit improvements

### PR DESCRIPTION
**Pull request by pavelsavara**
This whole PR is motivated by making `mono_wasm_exit` to be used.
- until now, it never run because the condition guarding it was always false
- we need it to run, so that MT build would stop all workers
- use `emscripten_force_exit`

Other fixes
- remember reason message of the abort
- refactor logging.ts, WS event handlers
- add native stdout and stdin flush
- test SecondRuntimeTest.js silently
- added one timer loop pass to `mono_run_main` so that queued tasks could finish running
- make all engines "NO_EXIT_RUNTIME=0" and balance it with `runtimeKeepalivePush`, `runtimeKeepalivePop`
- skip dryRun in test-main when we are not in secure context
- reorder `virtualWorkingDirectory` after memory snapshot
- refactor `Module` exit hooks into `register_exit_handlers` and `unregister_exit_handlers`

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]

File: src/mono/wasm/runtime/driver.c [hunk 1]
```diff
@@ -1186,6 +1186,8 @@ EMSCRIPTEN_KEEPALIVE int
 mono_wasm_exit (int exit_code)
 {
 	mono_jit_cleanup (root_domain);
+	fflush(stdout);
```
Comment (kg) [inline src/mono/wasm/runtime/driver.c:1189]: unimportant nit: space between fn name and (

**Comment by pavelsavara** [comment]
Is it possible that this is race condition on playwright rather than runtime ?

**Comment by pavelsavara** [comment]
More 
```
 Wasm.Build.Tests.HybridGlobalizationTests.AOT_HybridGlobalizationTests(buildArgs: BuildArgs { ProjectName = placeholder, Config = Release, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: Chrome, id: "afkbfwxy_pti") [FAIL]
      Assert.Contains() Failure: Sub-string not found
      String:    "[wasm test-browser] [8.0.0-prerelease.235"���
      Not found: "HybridGlobalization works, thrown excepti"���
      Stack Trace:
        /_/src/mono/wasm/Wasm.Build.Tests/HybridGlobalizationTests.cs(59,0): at Wasm.Build.Tests.HybridGlobalizationTests.TestHybridGlobalizationTests(BuildArgs buildArgs, RunHost host, String id, String extraProperties, Nullable`1 dotnetWasmFromRuntimePack)
        /_/src/mono/wasm/Wasm.Build.Tests/HybridGlobalizationTests.cs(29,0): at Wasm.Build.Tests.HybridGlobalizationTests.AOT_HybridGlobalizationTests(BuildArgs buildArgs, RunHost host, String id)
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
           at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
```
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-93472-merge-00e81291773748fb99/Workloads-Wasm.Build.Tests.HybridGlobalizationTests/1/console.cafc1503.log?helixlogtype=result)

```
Wasm.Build.Tests.PInvokeTableGeneratorTests.DllImportWithFunctionPointersCompilesWithWarning(buildArgs: BuildArgs { ProjectName = placeholder, Config = Release, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: Chrome, id: "l2sz5vlt_s2w") [FAIL]
      Assert.Contains() Failure: Sub-string not found
      String:    "[wasm test-browser] [8.0.0-prerelease.235"···
      Not found: "Main running"
      Stack Trace:
        /_/src/mono/wasm/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs(91,0): at Wasm.Build.Tests.PInvokeTableGeneratorTests.DllImportWithFunctionPointersCompilesWithWarning(BuildArgs buildArgs, RunHost host, String id)
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
           at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)
```
[Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-93472-merge-6a0b5754aa7140ce91/Workloads-Wasm.Build.Tests.PInvokeTableGeneratorTests/1/console.79ac2f41.log?helixlogtype=result)

File: src/mono/wasm/runtime/startup.ts [hunk 2]
```diff
@@ -234,8 +234,15 @@ async function onRuntimeInitializedAsync(userOnRuntimeInitialized: () => void) {
         await runtimeHelpers.afterPreRun.promise;
         mono_log_debug("onRuntimeInitialized");
 
-        runtimeHelpers.mono_wasm_exit = cwraps.mono_wasm_exit;
+        runtimeHelpers.mono_wasm_exit = (exit_code) => {
+            if (ENVIRONMENT_IS_NODE && !ENVIRONMENT_IS_WORKER) {
+                Module.runtimeKeepalivePop();
```
Comment (kg) [inline src/mono/wasm/runtime/startup.ts:239]: Is it wise to do this before calling mono_wasm_exit instead of after? Or in a finally block?

Comment (pavelsavara) [reply src/mono/wasm/runtime/startup.ts:239]: mono_wasm_exit on nodejs will call node's `exit()` and stop the node process when `runtimeKeepalivePop` balances. So it need be called before.

**Comment by radical** [comment]
> CI failures are #94221

Please merge that before this PR, so we can merge this one on green for WBT.

**Comment by radical** [comment]
> > CI failures are #94221
> 
> Please merge that before this PR, so we can merge this one on green for WBT.

Ignore that, since this is breaking on `main`.

**Review by maraf — APPROVED** [review]

**Comment by pavelsavara** [comment]
All CI failures are known

**Pull request merged**